File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/config/AmqpChannelFactory.java
package com.wks.servicemarketplace.accountservice.config;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.wks.servicemarketplace.accountservice.core.utils.CloseableUtils;
import org.glassfish.hk2.api.Factory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.io.IOException;

public class AmqpChannelFactory implements Factory<Channel> {

    private static final Logger LOGGER = LoggerFactory.getLogger(AmqpConnectionFactory.class);

    private Channel channel;

    @Inject
    public AmqpChannelFactory(Connection connection) throws IOException {
        this.channel = connection.createChannel();
    }

    @Override
    public Channel provide() {
        return channel;
    }

    @Override
    public void dispose(Channel instance) {
        CloseableUtils.close(instance, e -> LOGGER.error("Failed to dispose amqp channel", e));
    }
}

File: auth-service/src/main/kotlin/com/wks/servicemarketplace/authservice/config/AmqpChannelFactory.kt
package com.wks.servicemarketplace.authservice.config

import com.rabbitmq.client.BuiltinExchangeType
import com.rabbitmq.client.Channel
import com.rabbitmq.client.Connection
import com.wks.servicemarketplace.authservice.adapters.events.AutoDelete
import com.wks.servicemarketplace.authservice.adapters.events.Durable
import com.wks.servicemarketplace.authservice.adapters.events.Exchange
import com.wks.servicemarketplace.authservice.adapters.events.Internal
import com.wks.servicemarketplace.authservice.messaging.AuthMessaging
import org.glassfish.hk2.api.Factory
import javax.inject.Inject

class AmqpChannelFactory @Inject constructor(connection: Connection,
                                             parameters: ApplicationParameters) : Factory<Channel> {

    private val channel = connection.createChannel().also {
        it.basicQos(parameters.amqpPrefetchCount)
    }

    init {
        channel.exchangeDeclare(
                AuthMessaging.Exchange.MAIN,
                BuiltinExchangeType.TOPIC,
                Durable.TRUE,
                AutoDelete.FALSE,
                Internal.FALSE,
                emptyMap()
        )
    }

    override fun provide(): Channel = channel

    override fun dispose(instance: Channel?) {
        channel?.close()
    }
}